File: python/refactoring_alphabet/replaceconstructor.py
class CloudStorage:
    def __init__(self, name: str):
        # initialize a cloud resource, simulated by a file here
        self.name = name
        
    def write(self, data: bytes) -> None:
        # store data in cloud, an expensive operation
        f = open(self.name,"wb")
        f.write(data)
        f.close()
       
        
class Exporter:
    def __init__(self):
        # TODO: replace a direct constructor call with a factory call 
        self.storage = CloudStorage("export.csv")
        
    def export(self, data: bytes):
        self.storage.write(data)

# After start
class StorageFactory:
    def createStorage(self, name:str):
        return CloudStorage(name)
        
class ExporterwithFatoryCall:
    def __init__(self):
        self.storage = StorageFactory().createStorage("export.csv")
        
    def export(self, data: bytes):
        self.storage.write(data)
# After end

File: javascript/refactoring_alphabet/replaceconstructor.ts
import { writeFileSync } from 'fs';

// Simulating cloud storage with a file (imagine an actual cloud storage implementation)
class CloudStorage {
  constructor(public name: string) {}

  write(data: Uint8Array): void {
    writeFileSync(this.name, data);
  }
}

export class Exporter {
  storage: CloudStorage;

  constructor() {
    // TODO: replace a direct constructor call with a factory call
    this.storage = new CloudStorage('export.csv');
  }

  export(data: Uint8Array): void {
    this.storage.write(data);
  }
}

